### PR TITLE
Implement changes to fractracker mobile layer

### LINKED ISSUE #87: Add FracTracker mobile app data layer
FracTracker mobile app reports seem to be loaded via requests like this:

https://services.arcgis.com/jDGuO8tYggdCCnUJ/arcgis/rest/services/FracTrackerMobileAppNPCAMesaVerdeNationalPark_051416/FeatureServer/0/query?f=pbf&where=1%3D1&returnGeometry=true&spatialRel=esriSpatialRelIntersects&outFields=*&outSR=102100&resultOffset=0&resultRecordCount=8000&resultType=tile&quantizationParameters=%7B%22mode%22%3A%22edit%22%7D

Looks like arcgis FeatureLayer -- there's a plugin for them! https://esri.github.io/esri-leaflet/api-reference/layers/feature-layer.html

These are loaded under the Photo Documentation layer (2nd to last) here: http://ft.maps.arcgis.com/apps/MapJournal/index.html?appid=acb8b266e60446a2bf3fac099827c163

And are reported via the FracTracker app: https://play.google.com/store/apps/details?id=com.viableindustries.fractracker&hl=en

Response seems to be in "pbf" format... but changing `pbf` to `json` works:

https://services.arcgis.com/jDGuO8tYggdCCnUJ/arcgis/rest/services/FracTrackerMobileAppNPCAMesaVerdeNationalPark_051416/FeatureServer/0/query?f=json&where=1%3D1&returnGeometry=true&spatialRel=esriSpatialRelIntersects&outFields=*&outSR=102100&resultOffset=0&resultRecordCount=8000&resultType=tile&quantizationParameters=%7B%22mode%22%3A%22edit%22%7D

I think this should be pretty easy to add! @sagarpreet-chadha what do you think?

### PR DESCRIPTION
**Pull request by crisner**
Fixes #87  (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by crisner** [comment]
Work done:
- Implemented GeoJson layer that fetches data on map's moveend event

Thinks to do:
- Check and refactor popup code

127 0 0 1_5500_example_index html (5)

**Comment by crisner** [comment]
@jywarren, I have set the maximum number of results per page which is 250. I was thinking users would zoom in to get detailed information which would trigger a request resulting in loading markers they may have missed before. But some views may still have more than 250 results. How do you think we should implement this? Should I loop through the pages to load all the markers for the map view the user is on? The current layer setup won't allow the user to access the layer for zoom levels that are less than 5.

File: src/fracTrackerMobileLayer.js [hunk 1]
```diff
@@ -1,23 +1,101 @@
...
+        var northEast = bounds.getNorthEast();
+        var southWest = bounds.getSouthWest();
+        var left = southWest.lng;
+        var right = northEast.lng;
+        var top = northEast.lat;
+        var bottom = southWest.lat;
+        var polygon = left + ' ' + top + ',' + right + ' ' + top + ',' + right + ' ' + bottom + ',' + left + ' ' + bottom + ',' + left + ' ' + top;
+        
+        var $ = window.jQuery;
+        var fractrackerMobile_url = 'https://cors-anywhere.herokuapp.com/https://api.fractracker.org/v1/data/report?page=1&results_per_page=250&q={"filters":[{"name":"geometry","op":"intersects","val":"SRID=4326;POLYGON((' + polygon +'))"}],"order_by":[{"field":"report_date","direction":"desc"},{"field":"id","direction":"desc"}]}';
```
Comment (sagarpreet-chadha) [inline src/fracTrackerMobileLayer.js]: Hey,
What is this url cors-anywhere.herokuapp.com ?

**Comment by sagarpreet-chadha** [comment]
Hey @crisner ,

In the URL you have used: page=1&results_per_page=250
So this will fetch same 250 results everytime, right?

File: src/fracTrackerMobileLayer.js (hunk 1 above)
Comment (crisner) [reply src/fracTrackerMobileLayer.js]: This is to remove the cors policy issue. Jeff had suggested using it in the issue?

**Comment by crisner** [comment]
> In the URL you have used: page=1&results_per_page=250
So this will fetch same 250 results everytime, right?

It depends on your map view. But it would fetch the same results for the same map view. To put it simply it depends on the map's current bounds.

File: src/fracTrackerMobileLayer.js (hunk 1 above)
Comment (crisner) [reply src/fracTrackerMobileLayer.js]: I know it doesn't fetch the results if this is not prepended. But I am planning to look into it more tomorrow.

Comment (jywarren) [reply src/fracTrackerMobileLayer.js]: This allows us to fetch it despite CORS restrictions. We've confirmed this is OK with the FracTracker team. Thanks!

**Comment by jywarren** [comment]
I think this looks good to go! Shall we merge it?

**Comment by crisner** [comment]
I need to go over the popup information to make sure I haven't missed anything. It will be ready today (Jan 8). Thanks! :)

**Comment by crisner** [comment]
Also, noting it here to remind myself to update the leafletEnvironmentalLayers.js file with the change.

**Comment by crisner** [comment]
Changes done:
- Refactored popup code to handle `undefined` and `null` from data properties
- Changed layer name to match with the one listed in README.md
- Updated `AllLayers.js` file to reflect the changes
- Added promise to ajax request

**Comment by crisner** [comment]
I've also added a simple test to check if the layer gets added to the map. I tried to check for markers but because of the delay in response, the data doesn't become available when the test runs. There should be a better way to check this. I'll work on this later this week or when we are adding missing tests. For now, this PR is ready. Thanks!

**Comment by jywarren** [comment]
This is awesome! Thanks so much. Shall we bump the version number in package.json and also get the latest gh-pages branch pushed up so the main demo is live?

**Comment by crisner** [comment]
Sure. But if I'm not wrong, we had already bumped it once from `2.0.6` to `2.1.6` and hadn't published it. Do we bump it again?

**Comment by jywarren** [comment]
well, in theory only the commits up until the version change in package.json should be included, i think... i guess it's kind of an ambiguity though. I think I can just publish this as `2.1.6` since we ought to have published it as `2.1.0` before anyways. Sound good?

**Comment by crisner** [comment]
Yes, sounds good. :)

**Comment by jywarren** [comment]
publishing!!

**Comment by jywarren** [comment]
Sorry, do we need to do any `grunt build` on this before publishing? if so, i can just push out a 2.1.7?

**Comment by crisner** [comment]
Yes, I guess. We have git ignored the build file. So I think we have to run `grunt build` before publishing.

**Comment by crisner** [comment]
> if so, i can just push out a 2.1.7

Sounds good.

**Comment by jywarren** [comment]
ok, no prob - 2.1.7 coming up. I can release it fast enough that it'll supercede 2.1.6

**Comment by jywarren** [comment]
done!

**Comment by jywarren** [comment]
now doing gh-pages

**Comment by jywarren** [comment]
It's live! Let's test it out, then let's watch at `plots2` for dependabot to bump the LEL version in a PR there... i can give dependabot a nudge from its website console too...

**Comment by jywarren** [comment]
In the demo, how should I test this layer out? AHa! https://publiclab.github.io/leaflet-environmental-layers/example/#lat=41.531&lon=-96.542&zoom=6&layers=Standard,fracTrackerMobile,Power

**Comment by jywarren** [comment]
This is so great. I'm finding some things I'd like to make follow-ups for; is there a place to put these for now? 

* More layers (like FTM) in menu? - so before we publish, the current demo shows only a few of the layers... this wouldn't get copied into plots2 yet without more integration, right? I just don't want to publish on plots2 before all the existing layers are included. Thanks!
* Permalinks for popups? Like, if you open a popup maybe we can have a `#` button that links you back to this place AND it opens that same popup... this could be tough but useful... i don't see how we have a unique id here, but could we do it based on the layer and... timestamp? lat/lon? This might need some brainstorming.
* Less urgent: ways to show when a layer /will be active/ for the given bounding box - do we highlight it, do we have a tooltip to explain, do we show a bounding box when people hover on the layer?

This looks great!

https://publiclab.github.io/leaflet-environmental-layers/example/#lat=40.0333&lon=-75.1557&zoom=10&layers=Standard,fracTrackerMobile,skytruth,toxicReleaseLayer,mapknitter,Power

**Comment by jywarren** [comment]
image

So nice!

**Comment by crisner** [comment]
I am getting a `Uncaught ReferenceError: Spinner is not defined` when I try to access the above links.

**Comment by jywarren** [comment]
https://publiclab.github.io/leaflet-environmental-layers/example/#lat=40.1050&lon=-75.6771&zoom=14&layers=Standard,fracTrackerMobile,skytruth,toxicReleaseLayer,mapknitter,Power,Petroleum,Water

image

Some really nice side-by-side of reports and SkyTruth alerts for what's probably the same site here!

**Comment by jywarren** [comment]
Hmm. Would spinner be an npm req? I included all new npm modules, i think!

**Comment by crisner** [comment]
Yes. Looks like it is not included in the node modules. 

`GET https://publiclab.github.io/leaflet-environmental-layers/node_modules/leaflet-spin/example/spin/dist/spin.min.js net::ERR_ABORTED 404`

**Comment by jywarren** [comment]
I'll hold on bumping dependabot until we're sure here, so give me a heads up! Thanks for everything, this is so exciting :-)

Oh, one more idea! (sorry to dump them here, let's find a place to put them, and we don't have to act on all of these, they're just inspirations while I'm using it) -
    - So: We could have a url hash flag to have the menu expanded?

OK i'll try to include it in gh-pages and republish that.

**Comment by crisner** [comment]
Looks like blurred location and blurredlocationDisplay are also missing.

**Comment by jywarren** [comment]
Hm. That spin file is not present here: https://github.com/publiclab/leaflet-environmental-layers/tree/gh-pages/node_modules/leaflet-spin/example/spin

is it supposed to be part of the npm package as pulled in? Is it being gitignored?

**Comment by crisner** [comment]
> Oh, one more idea! (sorry to dump them here, let's find a place to put them, and we don't have to act on all of these, they're just inspirations while I'm using it) 

I have a few ideas myself. 😅  I was planning to note them all in a planning issue to discuss it first thing tomorrow morning. Maybe you could open a planning issue and put them there? and then I could add on to it?

>  We could have a url hash flag to have the menu expanded?

Awesome idea.

**Comment by crisner** [comment]
> Hm. That spin file is not present here: https://github.com/publiclab/leaflet-environmental-layers/tree/gh-pages/node_modules/leaflet-spin/example/spin
> 
> is it supposed to be part of the npm package as pulled in? Is it being gitignored?

Let me check

**Comment by jywarren** [comment]
OK awesome -- would you mind opening this list issue? That'd be super - and isn't our meeting Friday morning, not tomorrow?

**Comment by crisner** [comment]
Yes. Friday morning, your time.

**Comment by crisner** [comment]
> would you mind opening this list issue? 

Yes. Shall do it.

Btw, the dist file exists in the node modules in the local repo and node modules are git ignored in other branches anyway.

**Comment by crisner** [comment]
Sorry in the local machine. The file exists in my local copy.

**Comment by crisner** [comment]
And the file exists in the gh-pages branch of my copy https://github.com/crisner/leaflet-environmental-layers/tree/gh-pages/node_modules/leaflet-spin/example/spin/dist

The branch is not updated with the latest commits though. May be running `npm install` again would fix it?

**Comment by crisner** [comment]
I have opened an issue, #334 as a place to add and discuss new ideas. Shall edit the post with a few of my ideas and description tomorrow.

**Comment by jywarren** [comment]
You're right, I had missed some modules! Pushing them now.

**Comment by jywarren** [comment]
And regarding layers, it now appears to load all layers! Maybe it was the missing modules that stopped it.

**Comment by crisner** [comment]
I guess this resolves this [comment](https://github.com/publiclab/leaflet-environmental-layers/pull/320#issuecomment-572221079)? I'd like to note that unearthing layer is not included. Shall we include that too in the demo?

**Comment by jywarren** [comment]
Oh, that would be great. And yes! Thanks!

**Pull request merged**